### PR TITLE
Include pending silences for future muting decisions

### LINKED ISSUE #2426: Alert manager silence feature is working partially
Hi,

We came across an issue with our alert manager recently. We are able to create silences successfully and are also able to see the created silences in the UI. The issue is that the silence is working partially. 

Eg: Suppose there are 10 servers for a host with a label host_name. If I create a silence for the host_name, it should silence all alerts coming from all those 10 servers belonging to the host during the specified duration. What is happening is that sometimes 1 or 2 server alerts from the host are spilling into the slack channel. The rest majority of the alerts are silenced, but just a few of them at random are not silenced. Has this been observed previously in any version of alert manager? Was this a known issue in an older version of alert manager? 

Thanks,
Arnav

### PR DESCRIPTION
**Pull request by beorn7**
Previously, if a pending silence existed for an alert, and it later
became active without any silences getting added in the meantime, we
would miss the existence of that newly active silence.

This is my attempt to fix #2426 (and possibly others). This is a late night hack, and we should also add a unit test to expose the behavior. But I thought I put this out so that you can stop me if I'm on the wrong track here.

See #2586 for @waitstory's earlier attempt, which made me aware of the issue in the first place.

### CODE REVIEW

**Comment by beorn7** [comment]
@roidelapluie if we can finalize this in time for the ongoing 0.22 release, I can rebase this on top of the release branch.

**Comment by roidelapluie** [comment]
I plan to release 0.22 tomorrow. I am in favour of keeping that plan, and possibly release 0.22.1 next week with this.

**Comment by beorn7** [comment]
Yes, sure, but even then, I should rebase on top of the release branch.

But let's first see if this works at all.

**Comment by roidelapluie** [comment]
That, or I will cherry pick.

Le ven. 21 mai 2021 à 00:52, Björn Rabenstein ***@***.***> a
écrit :

**Comment by beorn7** [comment]
OK, I have added a test case to expose the bug and verify that this fixes it.

**Comment by beorn7** [comment]
I have also rebased this on top of the release-0.22 branch.

**Comment by roidelapluie** [comment]
Thank you, as I am not familiar with this code, I went for a .1 release with the API regression, because looking in details into this PR will require a bit more energy from me.

File: silence/silence.go
Comment (gotjosh) [inline silence/silence.go:175]: we no longer do any comparison of these so I don't think these need to sorted.

File: silence/silence.go
Comment (gotjosh) [inline silence/silence.go:163]:
```suggestion
   now := s.silences.now()
   level.Debug(s.logger).Log("msg", "determining silences state for deadline", "now", now, "count", len(allSils))
```

Comment (gotjosh) [reply silence/silence.go:163]: What do you think about this approach? I feel that the logline would be quite useful at aiding with debugging and would prefer a hard cutoff of this check to provide consistent results - I understand the difference would be quite minimal most of the time but seems quite practical.

**Review by gotjosh — COMMENTED** [review]
Great job @beorn7 👏  - I find this much simpler to understand than what we had before and have confirmed the fix works properly. I've left a few comments that are mostly nits, but please ping me again to have a second look.

File: silence/silence.go
Comment (beorn7) [reply silence/silence.go:159]: Or even "It is"?

File: silence/silence.go
Comment (beorn7) [reply silence/silence.go:163]: That's a very good point about consistent cutoff.

And the logline seems useful. I think I'll even extend it to include the number of active and pending IDs.

File: silence/silence.go
Comment (beorn7) [reply silence/silence.go:175]: But I think we are exposing the activeIDs via the API, and we shouldn't change the behavior here. pendingIDs are not exposed (they didn't exist before), but perhaps someone wants to add those later, so I thought I better sort them, too.

File: types/types.go
Comment (beorn7) [reply types/types.go:184]: I think that wouldn't work because `m.Status()` would try to acquire the mutex again.

File: types/types.go
Comment (beorn7) [reply types/types.go:209]: Yeah, generally, it's good Go design to allow `nil` instead of an empty slice. That's why I didn't set `pendingSilences`. But since `SilencedBy` and `InhibitedBy` are exported, I wasn't sure if I break any external assumptions, in particular because `AlertStatus` is marshaled as JSON, and I didn't want to take bets (or investigate in detail) if `nil` is marshaled in a different way as an empty string.

**Comment by beorn7** [comment]
I have addressed @gotjosh 's comments.

File: types/types.go
Comment (gotjosh) [reply types/types.go:209]: Makes sense.

File: types/types.go
Comment (gotjosh) [reply types/types.go:184]: A common pattern in Prometheus (which I tend to like) is that we do:

```
// Status implements Marker.
func (m *memMarker) Status(alert model.Fingerprint) AlertStatus {
	m.mtx.RLock()
	defer m.mtx.RUnlock()

	return m.status(alert)
}

func (m *memMarker) status(alert model.Fingerprint) AlertStatus {
	if s, found := m.m[alert]; found {
		return *s
	}
	return AlertStatus{
		State:       AlertStateUnprocessed,
		SilencedBy:  []string{},
		InhibitedBy: []string{},
	}
}
```

the suggestion is a bit silly but I find it quite important as it can be easy to miss when changing the signature of `Status`, and help reduce [shotgun surgery](https://en.wikipedia.org/wiki/Shotgun_surgery#:~:text=Shotgun%20surgery%20is%20an%20antipattern,implementations%20in%20a%20single%20change).

Comment (beorn7) [reply types/types.go:184]: Yes, that's a good point.

However, I would do such a refactoring separate from this PR as it is not really related.

Also, note that the `Status` method does something slightly different: It just returns an "empty" `AlertStatus` for a fingerprint that doesn't exist, _without adding it to the map_. It also returns a copy of the `AlertStatus` rather than a pointer. `SetInhibited` and `SetSilenced`, however, will modify the returned AlertStatus, so they need a pointer, and they need to add a newly created `AlertStatus` to the map.

I guess all of this can be somehow solved, but since it's getting pretty involved at this point, I am a bit concerned that we would break some assumptions we might not be aware of right now. I'm not super familiar with the code base after all. That's another reason to not do such a refactoring at the same time as a bugfix.

Comment (gotjosh) [reply types/types.go:184]: Perfectly sensible - let's defer for now.

**Review by gotjosh — APPROVED** [review]
LGTM.

**Comment by beorn7** [comment]
Thanks @gotjosh .

@roidelapluie / @simonpasquier should we wait for your review or do you think we can merge this now? It's a rather nasty bug that I would like to see fixed in a release ASAP.

**Review by roidelapluie — APPROVED** [review]
I have reviewed it over the weekend and LGTM

**Comment by beorn7** [comment]
\o/ Thank you very much.

Will you cut another release soon? I can also do it, if you are even busier than I am. (o:

File: silence/silence.go
Comment (roidelapluie) [reply silence/silence.go:183]: Note: this could go in a comment.

**Comment by vyshakrupeek** [comment]
Does this update fix this pagerduty issue :

https://github.com/prometheus/alertmanager/issues/635

**Pull request merged**
